[Step2]지하철 노선 기능 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다.

2단계 지하철 노선 기능 미션 구현이 완료되어, 코드 리뷰 요청드립니다.

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

re-mora committed:
docs: 요구사항 분석 및 구현 기능 목록 작성

re-mora committed:
refactor: 인수테스트 요청부의 파라미터 타입 변경

구현 편의 및 테스트 가독성을 위해 기존 Map을 이용한 파라미터 생성 방식에서, *Request객체의 생성자를 이용한 파라미터 생성 방식으로 변경
- AS-IS : Map
- TO-BE : T

파라미터 변경에 따른 코드 변경
- `StationRequest`에 모든 인자를 가지는 생성자 추가
- Map 관련 파라미터 생성 Utils 삭제
- Map을 이용한 RestAssured 요청 생성 부의 파라미터 타입 변경

인수테스트 검증 부 수정
- 검증 항목 가독성 향상을 위한 코드 정리

re-mora committed:
chore(re-package): domain별 package 구조 세분화

re-mora committed:
feat: 지하철 노선을 표현하는 Line Enttiy 및 Repository 구현

상행종점역과 하행종점역 사이의 지하철역 운행 경로를 표현하는 지하철 노선 Entity 정의

지하철 노선 Entity와 DB 연동을 위한 Repository Layer 정의 및 TC 작성

re-mora committed:
feat: 지하철 노선 생성 Service 및 TC 작성

- 요청의 상행/하행 종점역 번호 조회 및 조회 결과를 이용한 지하철 노선 생성
- 지하철 노선 생성에 필요한 요청/응답 객체 정의
- 구현기능 목록 갱신

re-mora committed:
feat: 지하철 노선 생성 실패 예외 TC 작성

- 존재 하지 않는 역 정보에 대한 예외 TC 작성
- 이미 존재하는 노선 정보에 대한 예외 TC 작성

re-mora committed:
feat: 지하철 노선 생성 API 구현

- 요청 수신 및 `201 Created` 응답 처리를 위한 Controller Layer 구현
- 지하철 노선 등록 기능에 대한 인수테스트 작성

re-mora committed:
feat: 지하철 노선 목록 조회 API 구현

- 지하철 노선 목록 조회 기능에 대한 인수테스트 작성
- TC내 반복적으로 작성되는 조건 생성 부 추출
- 전체 TC 수행 과정에서 데이터 중복으로 인해 일부 TC가 실패하는 현상 개선을 위해 각 TC 실행 시점에 테이블 초기화 처리 추가
- 기능 구현에 따른 docs 갱신

re-mora committed:
feat: 특정 지하철 노선 조회 API 구현

- 지하철 노선 조회 기능에 대한 인수테스트 작성
- TC내 반복적으로 작성되는 조건 생성 부 추출
- 기능 구현에 따른 docs 갱신

re-mora committed:
feat: 특정 지하철 노선 수정 API 구현

- 지하철 노선 수정 요청 객체 정의
- 지하철 노선 수정 기능에 대한 인수테스트 작성
- 기능 구현에 따른 docs 갱신

re-mora committed:
feat: 특정 지하철 노선 삭제 API 구현

- 지하철 노선 삭제 기능에 대한 인수테스트 작성
- 기능 구현에 따른 docs 갱신

re-mora committed:
refactor: 각 인수테스트의 독립된 수행을 위한 데이터 정리 부 구현

parkeeseul commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -1,32 +1,43 @@
+import org.springframework.test.context.TestConstructor;
+import org.springframework.test.context.TestConstructor.AutowireMode;
 
 @DisplayName("지하철역 관련 기능")
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@TestConstructor(autowireMode = AutowireMode.ALL)
+@Import(TearDownUtils.class)
 public class StationAcceptanceTest {
 
+    private final TearDownUtils tearDownUtils;
+
+    public StationAcceptanceTest(TearDownUtils tearDownUtils) {
+        this.tearDownUtils = tearDownUtils;
+    }
+
     @LocalServerPort
     int port;
```
새로운 인수테스트 클래스가 추가될 떄마다 비슷한 코드들이 반복될 것 같은데요
중복을 제거해 볼 수 있는 방법은 없을까요? 😄

re-mora replied:
각 인수테스트에 공통적으로 포함되는 어노테이션과 설정 부를 별도로 추출하여, 작성되는 각 인수테스트 클래스에 간략하게 주입 할 수 있을 것 같습니다. 👍 
미션 진행에 집중해서 설정 부 중복을 제거할 생각을 미처 하지 못했네요.. 
이 부분도 꼼꼼하게 확인해주셔서 감사합니다! :)

parkeeseul commented on src/test/java/nextstep/subway/station/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,175 @@
+        // Then
+        JsonPath jsonPath = getAllStationsResponse.jsonPath();
+        assertAll(
+            () -> assertCreated(createFirstLineResponse),
+            () -> assertCreated(createSecondLineResponse),
```
given 절을 검증하시는 이유가 있으실까요? 😄

re-mora replied:
습관적으로 테스트에 작성된 내용을 모두 검증부에 포함하여 작성하게 되었네요..
Given절의 `지하철 노선을 생성한다`에 해당하는 인수 테스트에서 해당 기능을 테스트하고 있으므로 삭제 해서 해당 인수테스트에 대한 검증이 더 눈에 잘보이도록 표현할 수 있을 것 같습니다.
이 부분도 꼼꼼하게 확인해 주셔서 감사합니다! :)

parkeeseul commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -35,56 +46,58 @@ public void setUp() {
-     * Then 지하철역 생성이 안된다
+     * Given 지하철역을 생성하고 When 기존에 존재하는 지하철역 이름으로 지하철역을 생성하면 Then 지하철역 생성이 안된다
      */
     @DisplayName("기존에 존재하는 지하철역 이름으로 지하철역을 생성한다.")
```
테스트 설명으로 보았을때 테스트의 의도를 파악하기 조금 힘들다고 생각되는데요,
`이미 존재하는 지하철역 이름을 지하철역을 생성할 수 없다.` 와 같이 적어보면 어떨까요? 😄

re-mora replied:
많은 것을 표현하려고 하다보니, 정작 표현력과 전달력이 좋지 않은 문장이 된 것 같아요 😭 
말씀하신 것처럼 이해하기 쉬운 표현으로 개선해보겠습니다! :)

parkeeseul commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -35,56 +46,58 @@ public void setUp() {
-        assertThat(stationNames)
-            .as("지하철역 목록 조회 시 생성한 지하철역 이름 포함 여부 검증")
-            .containsAnyOf(stationName);
+        Response getAllStationsResponse = get(STATION_BASE_URL).extract().response();
+        JsonPath jsonPath = getAllStationsResponse.jsonPath();
+        assertAll(
+            () -> assertCreated(stationCreateResponse),
+            () -> assertOk(getAllStationsResponse),
+            () -> assertThat(jsonPath.getList("name"))
+                .as("지하철역 목록 조회 시 생성한 지하철역 이름 포함 여부 검증")
+                .containsAnyOf(stationName)
+        );
     }
 
     /**
-     * Given 지하철역을 생성하고
-     * When 기존에 존재하는 지하철역 이름으로 지하철역을 생성하면
-     * Then 지하철역 생성이 안된다
+     * Given 지하철역을 생성하고 When 기존에 존재하는 지하철역 이름으로 지하철역을 생성하면 Then 지하철역 생성이 안된다
```
개행을 해주는 것이 가독성이 좋다고 생각하는데
어떻게 생각하시나요? 😄

re-mora replied:
미처 확인하지 못했는데, IDE의 코드 자동완성을 이용하다보니 주석 부분이 같이 정리된 것 같습니다.
각 단계를 개행으로 구분하여 인수 조건을 가독성 있게 표현 할 수 있도록 수정 하겠습니다. 
꼼꼼하게 확인해 주셔서 감사합니다 👍

parkeeseul commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,64 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = true)
```
👍

parkeeseul commented on src/test/java/nextstep/subway/domain/line/LineRepositoryTest.java:
```diff
@@ -0,0 +1,141 @@
+        );
+    }
+
+    @Test
+    @DisplayName("asdf")
+    public void asf() {
```
@_@ 테스트 메서드명과 테스트 설명을 확인해 주세요 😄

re-mora replied:
헉.. 급하게 작성하다 보니 임시로 작성한 부분을 한번 더 확인하지 못한것 같네요..
바로 수정하도록 하겠습니다 😭

parkeeseul reviewed (changes requested):
2단계 미션 잘 진행해 주셨네요! 👍 
몇가지 코멘트 남겨두었으니 확인 부탁드립니다! 😄 
즐거운 주말 되세요! 🔥

re-mora committed:
docs: 코드 리뷰 피드백 TODO List 작성

re-mora committed:
refactor: 인수테스트 작성 시 중복으로 작성되는 설정 부 추출

- AcceptanceTest : 인수테스트 작성에 공통적으로 명시되는 Annotations 설정을 Custom Annotation 으로 추출
- BaseTest : 인수 테스트 실행에 필요한 처리 부 추출(RestAssured port 설정, DB 데이터 초기화를 위한 Tear Down)

코드 구현에 따른 TODO List 갱신

re-mora committed:
chore(polish): LineAcceptanceTest.java

피드백 내용 적용에 따른 TODO List 갱신

re-mora committed:
chore(polish): Given절에 대한 검증 부 제거

이미 작성된 인수테스트에서 검증하고 있는 Given절에 대한 검증 부 제거
- Given절에 대한 검증부를 제거하여 해당 인수테스트에 대한 가독성을 향상시킬 수 있음

피드백 내용 적용에 따른 TODO List 갱신

re-mora commented:
안녕하세요 :) 
전반적으로 작성한 내용을 한번 더 돌아보지 못하고 PR 요청을 드려서 아쉬움이 많이 남습니다.. 😭 
리뷰해 주신 내용 확인 후 수정하였는데, 부족한 부분이 있다면 한번 더 피드백 부탁드리겠습니다 :)
즐거운 주말 보내세요 !

re-mora committed:
refactor: 지하철역 관련 인수테스트 작성 시 중복으로 작성되는 설정 부 추출

re-mora committed:
refactor: 인수조건의 각 단계를 별도의 한글로 표현된 메소드로 추출하여 가독성 개선

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요! 👍 
다음 미션 진행해 주세요! 😄 
즐거운 오후 되시고 화이팅 입니다 🔥
